### ISSUE TITLE
Allow pgmock to accept listener

### ISSUE THREAD

**Issue by kensaggy** [jackc/pgx#541]
First of all, great package - we use it extensively and really appreciate it :)

About the `pgmock` package and `Server` struct..
I understand that for your testing use-case you probably want whatever random port you can get, and it doesn't matter because you can get the full address using the exported `Addr()` func
But I have a use-case where I'd like to use the `pgmock` struct, however I'd like to be able to give it a listener - so that I can control the port deterministically.

I'd like to suggest either:
1. A new constructor function `NewServerWithListener` (top of mind name, open to suggestions) - or with the requested address string
2. Rename the `ln` member to exported as `Ln` (or `Listener`)
3. Add functional arguments to the `NewServer` so the signature will be something like `func NewServer(controller Controller, opts ...ServerOption)`

I'd be happy to submit a PR - just let me know if you agree with the idea and which approach you'll be more inclined to accept/approve

**Comment by jackc** [comment]
As you mentioned, at the moment the `pgmock` package is primarily designed for testing difficult to produce edge cases in pgx. But I am strongly considering is spinning off `pgmock` into its own package for the v4 major release of `pgx`. That would make it easier to use for external projects and would allow it to evolve quicker. I've already done that for `pgproto3` and `pgtype` among other v3 sub-packages.

I would have already done it but I have another half-finished project occupying the `github.com/jackc/pgmock` name. Among other things that project can act as a command line PostgreSQL proxy that logs messages. It's useful to build the scripts for `pgmock` by observing real behavior. So I'd kind of like to merge that tool and the current `pgx/pgmock` library. Just haven't done it yet.

With all that said, I think it's reasonable to add the functionality of passing in the listener as an argument into the current system. Not sure if option 1 or 3 would be better. The only caveat is that the package will almost certainly be extracted in the next few months and may undergo further interface changes at that time.

**Comment by jackc** [comment]
I've now extracted pgmock to https://github.com/jackc/pgmock. For the moment anyway, I removed the whole server concept. It needs easy timeout control. It is create deadlocks while testing. If/when it is going to be reintroduced it needs to have timeout control. I think your suggestions would fit with an overall restructuring of server. 

 For now the `Script` and `Step` interfaces can be used directly with an outside created server.
